drivers/mdio/mdio_xilinx_axienet.c

```diff
@@ -0,0 +1,200 @@
+
+static int mdio_xilinx_axienet_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+				    uint16_t *data)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *dev_data = dev->data;
+
+	enable_mdio_bus(config, dev_data);
+	xilinx_axienet_mdio_write_register(
+		config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET,
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE |
+			(prtad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR) |
+			(devad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR) |
+			XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READ);
+
+	mdio_bus_wait_ready(config);
+
+	*data = (uint16_t)(xilinx_axienet_read_mdio_register(
+				   config, XILINX_AXIENET_MDIO_READ_DATA_REG_OFFSET) &
+			   XILINX_AXIENET_MDIO_READ_DATA_REG_DATA_MASK);
+
+	LOG_DBG("Read %" PRIu16 " from MDIO!", *data);
+
+	return 0;
+}
+
+static int mdio_xilinx_axienet_write(const struct device *dev, uint8_t prtad, uint8_t devad,
+				     uint16_t data)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *dev_data = dev->data;
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_WRITE_DATA_REG_OFFSET, data);
+
+	enable_mdio_bus(config, dev_data);
+	xilinx_axienet_mdio_write_register(
+		config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET,
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE |
+			(prtad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR) |
+			(devad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR) |
+			XILINX_AXIENET_MDIO_CONTROL_REG_MASK_WRITE);
+
+	mdio_bus_wait_ready(config);
+
+	return 0;
+}
+
+static const struct mdio_driver_api mdio_xilinx_axienet_api = {
+	.bus_disable = mdio_xilinx_axienet_bus_disable,
+	.bus_enable = mdio_xilinx_axienet_bus_enable,
+	.read = mdio_xilinx_axienet_read,
+	.write = mdio_xilinx_axienet_write};
+
+#define XILINX_AXIENET_MDIO_INIT(inst)                                                             \
+                                                                                                   \
+	static const struct mdio_xilinx_axienet_config mdio_xilinx_axienet_config##inst = {        \
+		.reg = (void *)(uintptr_t)DT_INST_REG_ADDR(inst),                                  \
+		.clock_frequency_hz = DT_INST_PROP(inst, clock_frequency),                         \
+	};                                                                                         \
+	static struct mdio_xilinx_axienet_data mdio_xilinx_axienet_data##inst = {                  \
+		.bus_enabled = false,                                                              \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(inst, NULL, NULL, &mdio_xilinx_axienet_data##inst,                   \
+			      &mdio_xilinx_axienet_config##inst, POST_KERNEL,                      \
+			      CONFIG_MDIO_INIT_PRIORITY, &mdio_xilinx_axienet_api);
+
+#define DT_DRV_COMPAT xlnx_axi_ethernet_7_2_mdio
```

Comment:
this is okay but standard practice is to put at top of the file to make it easily recognizable, up to you though (I actually prefer this way in the PR but when I try to do it people don't seem to like it when they review)